xgboost_ray/sklearn.py

```diff
@@ -0,0 +1,901 @@
+        assert len(prediction.shape) <= 2
+        if len(prediction.shape) == 2 and prediction.shape[1] == n_classes:
+            return prediction
+        # binary logistic function
+        classone_probs = prediction
+        classzero_probs = 1.0 - classone_probs
+        return vstack((classzero_probs, classone_probs)).transpose()
+
+
+try:
+    from xgboost.sklearn import _is_cudf_df, _is_cudf_ser, _is_cupy_array
+except ImportError:
+    _is_cudf_df = None
+    _is_cudf_ser = None
+    _is_cupy_array = None
+
+try:
+    from xgboost.compat import XGBoostLabelEncoder
+except ImportError:
+    from sklearn.preprocessing import LabelEncoder as XGBoostLabelEncoder
+
+_RAY_PARAMS_DOC = """ray_params (Union[None, RayParams, Dict]): Parameters to configure
+            Ray-specific behavior. See :class:`RayParams` for a list of valid
+            configuration parameters. Will override ``n_jobs`` attribute
+            with own ``num_actors`` parameter.
+        _remote (bool): Whether to run the driver process in a remote
+            function. This is enabled by default in Ray client mode.
+
+"""
+
+_N_JOBS_DOC_REPLACE = (
+    """    n_jobs : int
+        Number of parallel threads used to run xgboost.  When used with other Scikit-Learn
+        algorithms like grid search, you may choose which algorithm to parallelize and
+        balance the threads.  Creating thread contention will significantly slow down both
+        algorithms.""",  # noqa: E501, W291
+    """    n_jobs : int
+        Number of Ray actors used to run xgboost in parallel.
+        In order to set number of threads per actor, pass a :class:`RayParams`
+        object to the relevant method as a ``ray_params`` argument. Will be
+        overriden by the ``num_actors`` parameter of ``ray_params`` argument
+        should it be passed to a method.""",  # noqa: E501, W291
+)
+
+
+def _treat_estimator_doc(doc: str) -> str:
+    """Helper function to make nececssary changes in estimator docstrings"""
+    doc = doc.replace(*_N_JOBS_DOC_REPLACE).replace(
+        "scikit-learn API for XGBoost",
+        "scikit-learn API for Ray-distributed XGBoost")
+    return doc
+
+
+def _xgboost_version_warn(f):
+    """Decorator to warn when xgboost version is < 1.4.0"""
+
+    @functools.wraps(f)
+    def inner_f(*args, **kwargs):
+        if XGBOOST_VERSION_TUPLE < (1, 4, 0):
+            warnings.warn(LEGACY_WARNING)
+        return f(*args, **kwargs)
+
+    return inner_f
+
+
+class RayXGBMixin:
+    """Mixin class to provide xgboost-ray functionality"""
+
+    def _ray_set_ray_params_n_jobs(
+            self, ray_params: Optional[Union[RayParams, dict]],
+            n_jobs: int) -> RayParams:
+        """Helper function to set num_actors in ray_params if not
+        set by the user"""
+        if ray_params is None:
+            if not n_jobs or n_jobs < 1:
+                n_jobs = 1
+            ray_params = RayParams(num_actors=n_jobs)
+        elif n_jobs:
```

Comment:
nit:

```suggestion
        elif n_jobs is not None:
```

And can we change the type hint from n_jobs to `Optional[int]` (which I think is accurate after looking at the sklearn code)?

Reply:
It is indeed more accurate